Source/WebCore/platform/graphics/mac/DisplayConfigurationMonitor.cpp

```diff
@@ -42,29 +42,65 @@ DisplayConfigurationMonitor::~DisplayConfigurationMonitor() = default;
 
 DisplayConfigurationMonitor& DisplayConfigurationMonitor::singleton()
 {
-    assertIsMainThread();
     static NeverDestroyed<DisplayConfigurationMonitor> instance;
     return instance;
 }
 
-void DisplayConfigurationMonitor::addClient(Client& client)
+void DisplayConfigurationMonitor::addClient(Client& client, SerialFunctionDispatcher* dispatcher)
 {
-    ASSERT(!m_clients.contains(&client));
-    m_clients.append(&client);
+    Locker locker { m_lock };
+    if (dispatcher)
+        assertIsCurrent(*dispatcher);
+    else
+        assertIsMainThread();
+    ASSERT(!m_clients.contains(ClientEntry { &client }));
+    m_clients.append({ &client, dispatcher });
 }
 
 void DisplayConfigurationMonitor::removeClient(Client& client)
 {
-    ASSERT(m_clients.contains(&client));
-    m_clients.removeFirst(&client);
+    Locker locker { m_lock };
+    size_t index = m_clients.find(ClientEntry { &client });
+    ASSERT(index != notFound);
+    if (m_clients.at(index).m_dispatcher)
+        assertIsCurrent(*m_clients.at(index).m_dispatcher);
+    else
+        assertIsMainThread();
+    m_clients.removeFirst(ClientEntry { &client });
 }
 
 void DisplayConfigurationMonitor::dispatchDisplayWasReconfigured(CGDisplayChangeSummaryFlags flags)
 {
     if (!(flags & kCGDisplaySetModeFlag))
         return;
-    for (auto* client : copyToVector(m_clients))
-        client->displayWasReconfigured();
+
+    Vector<ClientEntry> copy;
+    {
+        Locker locker { m_lock };
+        copy = copyToVector(m_clients);
+
+        // Dispatch notifications for Clients on other threads while we hold the lock, to
+        // ensure the dispatcher is still alive. Once on the other dispatcher, we'll
+        // re-check the validity of the client and then notify without the lock hold.
+        for (auto& [client, dispatcher] : copy) {
+            if (dispatcher) {
+                dispatcher->dispatch([this, client = client, dispatcher = dispatcher]() {
+                    {
+                        Locker locker { m_lock };
+                        if (!m_clients.contains(ClientEntry { client, dispatcher }))
+                            return;
```

Comment:
Your `operator==` doesn't look at the dispatcher, so you don't need to pass that in.

Up to you, but I wonder if it would be nicer to use `containsIf`, `findIf`, etc., to avoid needing an `operator==` that doesn't check both fields.

I think we can get into a situation where just before we take this lock, a `Client` with a given address is removed and destroyed, and then a new one with the same address is created and added. I suppose it's not a huge deal: the `displayWasReconfigured` call on the new `Client` is unnecessary, but isn't going to harm anything. Probably worth commenting about though. (If you wanted to protect against this I guess `Client` would need to have some unique identifier on it.)